### PR TITLE
backend: add Google Cloud storage support

### LINKED ISSUE #61: Add Google Cloud Storage Backend
Create a new backend implementation to support GCP Object storage

### PR DESCRIPTION
**Pull request by A-Hilaly**
Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

Fixes #61 

## Proposed Changes

- Add `cloud.google.com/go/storage` dependency
- Implement Cloud Storage backend 
- Add Cloud Storage to plugins config
- Edit flags/environment variables

### Checklist

- [x]  Read the CONTRIBUTING document. (It's checked since you are alread here.)
- [x]  Read the CODE OF CONDUCT document.
- [x]  Add tests to cover changes.
- [x]  Make sure your code follows the code style of this project
- [x]  Make sure CI and all other PR checks are green OR
- [x]  Code compiles correctly.
- [x]  Created tests which fail without the change (if possible).
- [x]  All new and existing tests passed.
- [x]  Add your changes to CHANGELOG.
- [x]  Improve and update the README (if necessary).
- [x]  Make sure documentation is up-to-date. Same file should also be updated in plugin index by sending a PR to that repository.

### CODE REVIEW

**Comment by kakkoyun** [comment]
@A-Hilaly This looks great so far. Thanks a lot for the contribution, this is one of the most wanted backends because GCP is very popular nowadays.

I'm sure you've noticed our contribution guidelines, we're also trying to improve those, PTAL #71. It'd be great if you could complete the PR checklist and rebase your branch.

And ping me whenever you need a review.

**Comment by A-Hilaly** [comment]
@kakkoyun  Done ! For the `CHANGELOG.md` file should i add bump the version and add changes to a [1.0.5] section or juste update the [1.0.4] one ?

**Comment by kakkoyun** [comment]
@A-Hilaly Thanks a lot. I'll have a look at it this weekend.
For the Changelog, could you please create an `Unreleased` titled section at the top and put your changes there. I'll handle the rest before release.

File: README.md [hunk 1]
```diff
@@ -141,10 +141,10 @@ GLOBAL OPTIONS:
    --skip-symlinks, --ss                       skip symbolic links in archive [$PLUGIN_SKIP_SYMLINKS, $ SKIP_SYMLINKS]
    --debug, -d                                 debug [$PLUGIN_DEBUG, $ DEBUG]
    --filesystem-cache-root value, --fcr value  local filesystem root directory for the filesystem cache (default: "/tmp/cache") [$PLUGIN_FILESYSTEM_CACHE_ROOT, $ FILESYSTEM_CACHE_ROOT]
-   --endpoint value, -e value                  endpoint for the s3 connection [$PLUGIN_ENDPOINT, $S3_ENDPOINT]
+   --endpoint value, -e value                  endpoint for the s3/cloud storage connection [$PLUGIN_ENDPOINT, $S3_ENDPOINT, $CLOUD_STORAGE_ENDPOINT]
    --access-key value, --akey value            AWS access key [$PLUGIN_ACCESS_KEY, $AWS_ACCESS_KEY_ID, $CACHE_AWS_ACCESS_KEY_ID]
-   --secret-key value, --skey value            AWS secret key [$PLUGIN_SECRET_KEY, $AWS_SECRET_ACCESS_KEY, $CACHE_AWS_SECRET_ACCESS_KEY]
-   --bucket value, --bckt value                AWS bucket name [$PLUGIN_BUCKET, $S3_BUCKET]
+   --secret-key value, --skey value            AWS/GCP secret key [$PLUGIN_SECRET_KEY, $AWS_SECRET_ACCESS_KEY, $CACHE_AWS_SECRET_ACCESS_KEY, $GCP_API_KEY]
```
Comment (kakkoyun) [inline README.md]: Overriding `bucket` and `endpoint` is ok, but I think we should create a dedicated parameter for GCP API key i.e `api-key`.

File: cache/backend/backend.go [hunk 2]
```diff
@@ -88,3 +89,30 @@ func InitializeFileSystemBackend(c FileSystemConfig, debug bool) (cache.Backend,
 
 	return newFileSystem(c.CacheRoot), nil
 }
+
+// CloudStorageConfig is a structure to store Cloud Storage backend configuration
+type CloudStorageConfig struct {
+	Bucket     string
+	ACL        string
```
Comment (kakkoyun) [inline cache/backend/backend.go]: I don't think ACL is used in anywhere, let's remove it.

Comment (kakkoyun) [reply cache/backend/backend.go]: Or use it, if we have an option to pass it.

Comment (kakkoyun) [reply cache/backend/backend.go]: https://godoc.org/cloud.google.com/go/storage#ACLHandle.Set

File: cache/backend/cloud_storage.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// Get returns an io.Reader for reading the contents of the file
+func (c *cloudStorageBackend) Get(p string) (io.ReadCloser, error) {
+	bkt := c.client.Bucket(c.bucket)
+	obj := bkt.Object(p)
+	if c.encryption != "" {
+		obj = obj.Key([]byte(c.encryption))
+	}
+
+	// TODO: use a timeout or cancel context
+	return obj.NewReader(context.TODO())
```
Comment (kakkoyun) [inline cache/backend/cloud_storage.go]: Let's use `context.Background()` for now. We should introduce a `timeout` value for operations. I'll open an issue for it.

File: cache/backend/cloud_storage.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// Put uploads the contents of the io.ReadSeeker
+func (c *cloudStorageBackend) Put(p string, src io.ReadSeeker) error {
+	bkt := c.client.Bucket(c.bucket)
+	obj := bkt.Object(p)
+	if c.encryption != "" {
+		obj = obj.Key([]byte(c.encryption))
+	}
+
+	//TODO: use a timeout or cancel context
+	w := obj.NewWriter(context.TODO())
```
Comment (kakkoyun) [inline cache/backend/cloud_storage.go]: Let's use `context.Bacground()` for now.

File: cache/backend/backend.go (hunk 2 above)
Comment (A-Hilaly) [reply cache/backend/backend.go]: Ohhh thanks ! i didn't find how to set ACL last time i tried

**Comment by A-Hilaly** [comment]
@kakkoyun Thanks for the review, i made the changes as requested. I still have the e2e tests to write. Will need to understand how they are working in order to add the correct tests.

**Comment by kakkoyun** [comment]
Hey @A-Hilaly, since you worked on GCP and aware of the AWS integration, I would like to have your opinion on https://github.com/google/go-cloud. This project is an abstraction over AWS, GCP and Azure. Maybe we can use this for object store integrations and provide integrations for other vendors like Azure as well. I'm not suggesting anything, just food for thought.

**Comment by A-Hilaly** [comment]
I think that's a very good idea ! it will be a tree birds one stone deal 😄 i also like the fact that it uses only the standard sdk for each cloud nothing more https://github.com/google/go-cloud/blob/master/go.mod
drone-cache will also benefit whenever go-cloud implement support a new blop storage provider.

**Comment by kakkoyun** [comment]
Thanks for the comment. Good point I'll check it.
Let's finish up with this PR since we are nearly at the end of it.
I have created an issue for go-cloud #76.

**Comment by kakkoyun** [comment]
@A-Hilaly Would you be able to push this through? Let me know if you need any help.

**Comment by A-Hilaly** [comment]
@A-Hilaly Yes i'm working on this PR now. Need to rebase and add the plugin tests. Not sure how to use APIKey based access with minio (S3 uses SecretKey and AccessKey)

**Comment by kakkoyun** [comment]
@A-Hilaly Cool thanks! I'm not sure if minio could be used to test google cloud storage locally, we need to check. If not there must be a someway to emulate google cloud storage, we should just find it.

File: go.mod [hunk 5]
```diff
@@ -1,22 +1,23 @@
...
 	github.com/smartystreets/assertions v0.0.0-20190116191733-b6c0e53d7304 // indirect
 	github.com/smartystreets/goconvey v0.0.0-20181108003508-044398e4856c // indirect
 	github.com/urfave/cli v1.20.0
 	golang.org/x/crypto v0.0.0-20190820162420-60c769a6c586
+	golang.org/x/sys v0.0.0-20191008105621-543471e840be // indirect
+	google.golang.org/api v0.9.0
 	gopkg.in/ini.v1 v1.41.0 // indirect
 )
+
+go 1.13
```
Comment (A-Hilaly) [inline go.mod]: @kakkoyun is it fine to keep this line ? i don't know if this breaks compilation using older versions

Comment (kakkoyun) [reply go.mod]: We should remove it. I'll send another dedicated PR to upgrade everything to 1.13
We have to upgrade container base images, CI images etc.

**Comment by A-Hilaly** [comment]
i made some research around GCS testing technics and i have found these two interesting projects:
- fake CS server we can run as a docker container (https://github.com/fsouza/fake-gcs-server) 
- a mock library https://github.com/googleapis/google-cloud-go-testing/tree/master/storage/stiface

I personally like the fake gcs server it could be very useful to write e2e tests. What do you think @kakkoyun ?

**Comment by kakkoyun** [comment]
@A-Hilaly I'm %100 with you on this one, just go with e2e tests. We only have e2e tests right now anyway.
`fake-gcs-server` it is 👍

**Comment by A-Hilaly** [comment]
@kakkoyun i tried to add the end to end tests, bucket creation and object creation works as expected but reading object fails with a weird error:

```bash
~/source/drone-cache(gcs-support*) » go test -v -timeout 30s github.com/meltwater/drone-cache/cache/backend -run ^TestCloudStorageTruth
=== RUN   TestCloudStorageTruth
--- FAIL: TestCloudStorageTruth (0.71s)
    cloud_storage_test.go:66: googleapi: got HTTP response code 403 with body: <?xml version='1.0' encoding='UTF-8'?><Error><Code>UserProjectAccountProblem</Code><Message>User project billing account not in good standing.</Message><Details>The billing account for project 47866975936 is disabled in state absent</Details></Error>
FAIL
FAIL    github.com/meltwater/drone-cache/cache/backend  0.860s
FAIL
```

i'll try to investigate more this error with different `fake-gcs-servers` versions

**Comment by kakkoyun** [comment]
Hey @A-Hilaly, I will try to give this one a push this weekend or in the next week. 

What's left? Do you need help with something specific?

**Comment by A-Hilaly** [comment]
Hey @kakkoyun, i couldn't have all the end2end test fixed. Specifically this part https://github.com/A-Hilaly/drone-cache/blob/gcs-support/cache/backend/cloud_storage_test.go#L69-L73 . I'm not sure why this happens, i'm suspecting the fake-gcs-server.

**Comment by kakkoyun** [comment]
@A-Hilaly I have rebased it, and I'm going to merge it. 
I'll address testing issues in #45 before we cut `v1.1.0` release.

**Pull request merged**